[Test] Creating ACL 2 6 Python3 Test Module

j-ororke opened the pull request:
Creating ACL_2_6 python3 test module:
- Following test steps mentioned in test plan PR #[5061](https://github.com/CHIP-Specifications/chip-test-plans/pull/5061) from Amine
- Using base branch of PR #[38263](https://github.com/project-chip/connectedhomeip/pull/38263), once that PR has merged to master, then this PR can come out of WIP.

#### Testing
- WSL - Windows Subsystem Linux environment

Alami-Amine committed:
Modifying how ACL Attribute is written: making it use a non-empty initial ReplaceAll list

Alami-Amine committed:
removing Test modifications on ACL Extensions

Alami-Amine committed:
adding reference to issue in TODO comment

Alami-Amine committed:
integrating comments

Alami-Amine committed:
integrating comments

Alami-Amine committed:
Revert "removing Test modifications on ACL Extensions"

This reverts commit 055af80d6c71e40ddeef081b6046263bc06ef141.

Alami-Amine committed:
Clarifying AttributeDataIB Checkpoint

Alami-Amine committed:
adding Valid CAT values

Alami-Amine committed:
Integrating comments

Alami-Amine committed:
Add comments clarifying AttributeDataIB rollback

Alami-Amine committed:
Integrating comments again

Alami-Amine committed:
change method order

j-ororke committed:
Creating ACL_2_6 python3 test module:
- Following test steps mentioned in test plan PR #[5061](https://github.com/CHIP-Specifications/chip-test-plans/pull/5061) from Amine
- Using base branch of PR #[38263](https://github.com/project-chip/connectedhomeip/pull/38263)

restyled-commits committed:
Restyled by autopep8

restyled-commits committed:
Restyled by isort

j-ororke committed:
Updating TC_ACL_2_6 python3 test module:
- Resolving linting errors
- Added additional checks for test step 3 and 5 to include struct validation

Alami-Amine commented on src/python_testing/TC_ACL_2_6.py:
```diff
@@ -0,0 +1,247 @@
+            self.print_step("Examining event", str(event_data))
+
+            if hasattr(event_data, 'Data') and hasattr(event_data.Data, 'changeType'):
+                # Check for initial admin entry (first event)
```
interesting, so you choose to just list all events since the beginning of the test and check them, instead of using Event Filtering, right?
- because the current YAML filters the events (using `eventNumber: "LastReceivedEventNumber + 1`), that way we only see the events of that Test Step
- Also, in the TestPlan (although they don't mention filtering), they say  "AccessControlEntryChanged eventscontaining at least 3 (after my change its 2) new elements" --> which hints to event filtering

j-ororke replied:
I have updated test step 5 to match what we discussed in slack messages, thank you for the clarification and update to this test step. Please let me know if this conforms to the updated test plan.

j-ororke committed:
Updated TC_ACL_2_6 python test module:
- Rebased branch with upstream master
- Changed to checking for 2 new elements during test step 5
- Changed print_step() to logging.info()

restyled-commits committed:
Restyled by autopep8

j-ororke committed:
Updating to the upstream master branch version of WriteClient.cpp, WriteClient.h, Test_TC_ACL_2_5.yaml, and TestWriteChunking.cpp, to remove changes included in commit on this branch

j-ororke committed:
Update src/python_testing/TC_ACL_2_6.py

Co-authored-by: Amine Alami <43780877+Alami-Amine@users.noreply.github.com>

j-ororke committed:
Removing Test_TC_ACL_2_6 yaml script

j-ororke committed:
Updating TC_ACL_2_6 python3 test module:
- Updating endpoint to 0 for REPL Linux CI checks to pass as AccessControl cluster events are on endpoint 0

cecille commented on src/python_testing/TC_ACL_2_6.py:
```diff
@@ -0,0 +1,237 @@
+        asserts.assert_true(found_initial_event, "Did not find expected initial ACL event")
+
+        # Add a small delay before reading events again
+        await asyncio.sleep(5)
```
what's this for? Why would the event list change between these two reads?

j-ororke replied:
Ah, sorry there was an issue in the CI, if we use endpoint = 1 then there are no events noticed as AccessCluster events appear to be on endpoint 0, so during debugging to find out why CI was failing, I had initially thought it might be a timing issue, I have removed this from the code here now.

j-ororke commented on src/python_testing/TC_ACL_2_6.py:
```diff
@@ -0,0 +1,237 @@
+        f1 = await self.read_single_attribute_check_success(endpoint=0, cluster=oc_cluster, attribute=cfi_attribute)
+
+        self.step(3)
+        acec_event = Clusters.AccessControl.Events.AccessControlEntryChanged
```
No problem, created a short follow-up task here: https://github.com/project-chip/matter-test-scripts/issues/548

j-ororke committed:
Updating TC_ACL_2_6 python3 test module:
- Updating to using a simple for loop to validate the ReadEvent() response data
- Removed unneeded asyncio.wait() from prior debugging
- Created a short follow-up task PR to add subscription to this test module later

j-ororke commented on src/python_testing/TC_ACL_2_6.py:
```diff
@@ -0,0 +1,237 @@
+        for event_data in events_response:
+            logging.info(f"Examining initial event {str(event_data)}")
+
+            if hasattr(event_data, 'Data') and hasattr(event_data.Data, 'changeType'):
```
So. sadly I had to validate a bit differently than was able to in ACL_2_4, given that the structs are a bit different with the validated structs in 2_4 being for attributes and the structs in this test are for events. 
Thus it was a bit different, because I really liked the one liner asserts.assert_in() that we were able to use in 2.4, sadly we could not use it here without modifying the response structs and then it became more code than before. 
However, I was able to just do a simple for loop to clean the validation up in this test while maintaining the validation we needed and eliminating a few lines of code, I hope making the code a lil bit cleaner.

j-ororke committed:
Update TC_ACL_2_6.py

Resolving linting errros

j-ororke committed:
Merge branch 'master' into update_ACL_2_6_test
